editor/src/clj/editor/future.clj

```diff
@@ -29,14 +29,36 @@
   ^CompletableFuture [x]
   (CompletableFuture/completedFuture x))
 
-(defmacro supply-async [& body]
+(defmacro compute
+  "Asynchronously perform a CPU-intensive operation (see also: io)"
+  [& body]
   `(let [bindings# (get-thread-bindings)]
      (CompletableFuture/supplyAsync
        (reify Supplier
          (get [~'_]
            (with-bindings bindings#
              ~@body))))))
 
+(def io-executor
+  (let [counter (atom 0)]
+    (Executors/newCachedThreadPool
+      (reify ThreadFactory
+        (newThread [_ r]
+          (doto (Thread. r)
+            (.setDaemon true)
+            (.setName (str "editor.future/io-executor#" (swap! counter inc)))))))))
+
+(defmacro io
+  "Asynchronously perform an IO-intensive operation (see also: compute)"
+  [& body]
+  `(let [bindings# (get-thread-bindings)]
+     (CompletableFuture/supplyAsync
+       (reify Supplier
+         (get [~'_]
+           (with-bindings bindings#
+             ~@body)))
+       io-executor)))
+
```

Comment:
Should these respect `editor.disk-availability/avaliable?` and block until for instance, a save operation completes?

Reply:
No, this is in no way related to editor disc availability. `io` is used for blocking tasks, for example, for file system access or for piping a process output of a process launched by the editor script.